downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,2028 @@
+        nix::sys::uio::pwrite(file.as_raw_fd(), &[1u8], offset as i64)
+            .map_err(|e| {
+                CrucibleError::IoError(format!(
+                    "writing dirty byte failed: {e}",
+                ))
+            })?;
+        Ok(())
+    }
+
+    /// Returns the total size of the raw data file
+    ///
+    /// This includes block data, context slots, active slot array, and metadata
+    fn file_size(&self) -> u64 {
+        let block_count = self.block_count();
+        self.block_size().checked_mul(block_count).unwrap()
+            + BLOCK_META_SIZE_BYTES
+            + (block_count + 7) / 8
+            + BLOCK_CONTEXT_SLOT_SIZE_BYTES * block_count * 2
+    }
+
+    /// Returns the beginning of supplementary data in the file
+    fn supplementary_data_offset(&self) -> u64 {
+        self.block_count() * self.block_size()
+    }
+
+    /// Returns the byte offset of the given context slot in the file
+    ///
+    /// Contexts slots are located after block data in the extent file. There
+    /// are two context slots arrays, each of which contains one context slot
+    /// per block.  We use a ping-pong strategy to ensure that one of them is
+    /// always valid (i.e. matching the data in the file).
+    fn context_slot_offset(&self, block: u64, slot: ContextSlot) -> u64 {
+        self.supplementary_data_offset()
+            + (self.block_count() * slot as u64 + block)
+                * BLOCK_CONTEXT_SLOT_SIZE_BYTES
+    }
+
+    /// Number of blocks in the extent file
+    fn block_count(&self) -> u64 {
+        self.extent_size.value
+    }
+
+    /// Returns the byte offset of the `active_context` bitpacked array
+    fn active_context_offset(&self) -> u64 {
+        self.supplementary_data_offset()
+            + self.block_count() * 2 * BLOCK_CONTEXT_SLOT_SIZE_BYTES
+    }
+
+    fn active_context_size(&self) -> u64 {
+        (self.block_count() + 7) / 8
+    }
+
+    fn metadata_offset(&self) -> u64 {
+        self.active_context_offset() + self.active_context_size()
+    }
+
+    /// Number of bytes in each block
+    fn block_size(&self) -> u64 {
+        self.extent_size.block_size_in_bytes() as u64
+    }
+
+    fn get_metadata(&self, file: &File) -> Result<OnDiskMeta, CrucibleError> {
+        let mut buf = [0u8; BLOCK_META_SIZE_BYTES as usize];
+        let offset = self.metadata_offset();
+        nix::sys::uio::pread(file.as_raw_fd(), &mut buf, offset as i64)
+            .map_err(|e| {
+                CrucibleError::IoError(format!("reading metadata failed: {e}"))
+            })?;
+        let out: OnDiskMeta = bincode::deserialize(&buf)
+            .map_err(|e| CrucibleError::BadMetadata(e.to_string()))?;
+        Ok(out)
+    }
+
+    fn write_context_slots_contiguous<'a, I>(
+        &self,
+        file: &File,
+        block_start: u64,
+        iter: I,
+        slot: ContextSlot,
+    ) -> Result<(), CrucibleError>
+    where
+        I: Iterator<Item = Option<&'a DownstairsBlockContext>>,
```

Comment:
I experimented with changing this signature, with mixed results.  It turns out that we use the possibility of writing `None` in a few places:

- When migrating instances, to write the context slots in bulk
- When defragmenting context slots, for the same reason

It's possible to work around this, but it makes this code much trickier (see the diff below).

Because this is a private function – no one outside this module can call it – I'd vote to leave it as is, but would reconsider if you feel strongly about it.

For reference, here's the diff:

```diff
diff --git a/downstairs/src/extent_inner_raw.rs b/downstairs/src/extent_inner_raw.rs
index b8459d8..6d8a624 100644
--- a/downstairs/src/extent_inner_raw.rs
+++ b/downstairs/src/extent_inner_raw.rs
@@ -545,18 +545,22 @@ impl RawInner {
         assert_eq!(block_count, ctxs.len());
 
         file.set_len(layout.file_size())?;
-        layout.write_context_slots_contiguous(
-            file,
-            0,
-            ctxs.iter().map(Option::as_ref),
-            ContextSlot::A,
-        )?;
-        layout.write_context_slots_contiguous(
-            file,
-            0,
-            std::iter::repeat(None).take(block_count),
-            ContextSlot::B,
-        )?;
+        for chunk in ctxs
+            .iter()
+            .group_by(|b| b.is_some())
+            .into_iter()
+            .filter(|(some, _group)| *some)
+            .map(|(_some, group)| group)
+        {
+            layout.write_context_slots_contiguous(
+                file,
+                0,
+                chunk.map(|b| b.as_ref().unwrap()),
+                ContextSlot::A,
+            )?;
+        }
+        // `ContextSlot::B` is already initialized to all zeros, which
+        // deserializes to `None`.
 
         layout.write_active_context_and_metadata(
             file,
@@ -914,10 +918,7 @@ impl RawInner {
             let mut group = group.peekable();
             let start = group.peek().unwrap().block;
             self.layout.write_context_slots_contiguous(
-                &self.file,
-                start,
-                group.map(Option::Some),
-                slot,
+                &self.file, start, group, slot,
             )?;
             writes += 1;
         }
@@ -1113,12 +1114,25 @@ impl RawInner {
                     self.sync_index + 1;
             }
         }
-        let r = self.layout.write_context_slots_contiguous(
-            &self.file,
-            counter.min_block,
-            dest_slots.iter().map(|v| v.as_ref()),
-            !copy_from,
-        );
+
+        let mut r = Ok(());
+        for chunk in dest_slots
+            .iter()
+            .group_by(|b| b.is_some())
+            .into_iter()
+            .filter(|(some, _group)| *some)
+            .map(|(_some, group)| group)
+        {
+            if let Err(e) = self.layout.write_context_slots_contiguous(
+                &self.file,
+                0,
+                chunk.map(|b| b.as_ref().unwrap()),
+                !copy_from,
+            ) {
+                r = Err(e);
+                break;
+            }
+        }
 
         // If this write failed, then try recomputing every context slot
         // within the unknown range
@@ -1251,7 +1265,7 @@ impl RawLayout {
         slot: ContextSlot,
     ) -> Result<(), CrucibleError>
     where
-        I: Iterator<Item = Option<&'a DownstairsBlockContext>>,
+        I: Iterator<Item = &'a DownstairsBlockContext>,
     {
         let mut buf = self.buf.take();
         buf.clear();
@@ -1259,11 +1273,11 @@ impl RawLayout {
         for block_context in iter {
             let n = buf.len();
             buf.extend([0u8; BLOCK_CONTEXT_SLOT_SIZE_BYTES as usize]);
-            let d = block_context.map(|b| OnDiskDownstairsBlockContext {
-                block_context: b.block_context,
-                on_disk_hash: b.on_disk_hash,
-            });
-            bincode::serialize_into(&mut buf[n..], &d).unwrap();
+            let d = OnDiskDownstairsBlockContext {
+                block_context: block_context.block_context,
+                on_disk_hash: block_context.on_disk_hash,
+            };
+            bincode::serialize_into(&mut buf[n..], &Some(d)).unwrap();
         }
         let offset = self.context_slot_offset(block_start, slot);
         nix::sys::uio::pwrite(file.as_raw_fd(), &buf, offset as i64).map_err(
```

Reply:
Circling back here, are those two cases still true?

- When migrating, the context slots start as all zeroes, which deserialize as None.
- After defragmenting, both context slots contain the same data - the current code doesn't assign None to overwrite one of them.